File: backend/geranet.py
"""Integração com a API Geranet para consulta e emissão de NFSe no padrão nacional."""

import base64
import logging
from typing import Optional
from pathlib import Path

import requests
from config import settings

logger = logging.getLogger(__name__)

GERANET_BASE_URL = "https://nfe.geranet.net/api/v1"


def arquivo_para_hexadecimal(caminho_arquivo: str) -> str:
    """
    Lê um certificado digital A1 (.pfx) e retorna seu conteúdo em hexadecimal.

    Args:
        caminho_arquivo: Caminho absoluto para o arquivo .pfx

    Returns:
        String hexadecimal do conteúdo do certificado
    """
    return Path(caminho_arquivo).read_bytes().hex()


def _obter_certificado_hex() -> str:
    """
    Obtém o certificado digital em hexadecimal automaticamente.

    Prioridade:
    1. GERANET_CERT_BASE64 (env) — útil no Railway
    2. GERANET_CERT_PATH (caminho do .pfx) — uso local

    Returns:
        String hexadecimal do certificado

    Raises:
        ValueError: Se nenhuma fonte estiver configurada
    """
    if settings.GERANET_CERT_BASE64:
        logger.info("Carregando certificado via GERANET_CERT_BASE64")
        return base64.b64decode(settings.GERANET_CERT_BASE64).hex()

    if settings.GERANET_CERT_PATH:
        caminho = settings.GERANET_CERT_PATH
        if not Path(caminho).exists():
            raise ValueError(f"Certificado não encontrado: {caminho}")
        logger.info("Carregando certificado de: %s", caminho)
        return Path(caminho).read_bytes().hex()

    raise ValueError(
        "Nenhum certificado configurado. Defina GERANET_CERT_BASE64 ou "
        "GERANET_CERT_PATH no .env"
    )


def _headers() -> dict:
    """Headers padrão para requisições à Geranet."""
    return {
        "Authorization": f"Bearer {settings.GERANET_API_KEY}",
        "Content-Type": "application/json",
    }


def consultar_notas(
    cnpj: str,
    inscricao_municipal: str,
    razao_social: str,
    municipio: str,
    certificado_digital: Optional[str] = None,
    senha_certificado: Optional[str] = None,
    ultimo_nsu: str = "0",
    chave_nfse: Optional[str] = None,
    timeout: int = 30,
) -> dict:
    """
    Consulta NFSe no portal nacional via Geranet.

    POST /api/v1/nfse/consultar-notas

    Args:
        cnpj: CNPJ do prestador (apenas números)
        inscricao_municipal: Inscrição municipal do prestador
        razao_social: Razão social do prestador
        municipio: Código IBGE do município (7 dígitos)
        certificado_digital: Conteúdo do certificado A1 em hexadecimal.
                            Se None, carrega automaticamente das configs.
        senha_certificado: Senha do certificado A1.
                          Se None, usa GERANET_CERT_PASSWORD do .env.
        ultimo_nsu: NSU para paginação (0 para começar)
        chave_nfse: Chave DF-e específica para consulta direta (opcional)
        timeout: Timeout em segundos

    Returns:
        Dict com a resposta completa da API

    Raises:
        ConnectionError: Se não conseguir conectar à API
        ValueError: Se a API retornar erro
    """
    if not settings.GERANET_API_KEY:
        raise ValueError("GERANET_API_KEY não configurada no .env")

    # Carrega certificado automaticamente se não fornecido
    hex_cert = certificado_digital if certificado_digital else _obter_certificado_hex()
    senha = senha_certificado if senha_certificado else settings.GERANET_CERT_PASSWORD

    if not senha:
        raise ValueError(
            "Senha do certificado não informada. Envie 'senha_certificado' ou "
            "configure GERANET_CERT_PASSWORD no .env"
        )

    payload = {
        "prestador": {
            "cnpj": cnpj,
            "inscricaoMunicipal": inscricao_municipal,
            "razaoSocial": razao_social,
            "municipio": municipio,
        },
        "certificadoDigital": hex_cert,
        "senhaCertificadoDigital": senha,
        "padraoNacional": "sim",
        "ultimoNsu": ultimo_nsu,
    }

    if chave_nfse:
        payload["chaveNfse"] = chave_nfse

    try:
        logger.info(
            "Consultando NFSe via Geranet: CNPJ=%s, Mun=%s, NSU=%s",
            cnpj, municipio, ultimo_nsu,
        )
        resp = requests.post(
            f"{GERANET_BASE_URL}/nfse/consultar-notas",
            headers=_headers(),
            json=payload,
            timeout=timeout,
        )
        logger.info("Geranet respondeu HTTP %s", resp.status_code)

        if resp.status_code == 422:
            erro = resp.json()
            detalhes = erro.get("detail", [])
            msgs = "; ".join(
                e.get("msg", str(e)) for e in (detalhes if isinstance(detalhes, list) else [detalhes])
            )
            raise ValueError(f"Dados inválidos: {msgs}")

        if resp.status_code == 401:
            raise ValueError("Credenciais Geranet inválidas (HTTP 401)")

        if resp.status_code == 402:
            raise ValueError("Saldo insuficiente na Geranet (HTTP 402)")

        if resp.status_code >= 400:
            raise ValueError(
                f"Geranet retornou HTTP {resp.status_code}: {resp.text[:300]}"
            )

        dados = resp.json()

        if dados.get("situacao") == "erro":
            raise ValueError(f"Geranet retornou erro: {dados.get('mensagem', 'sem mensagem')}")

        return dados

    except requests.Timeout:
        raise ConnectionError(f"Geranet não respondeu em {timeout}s (timeout)")
    except requests.ConnectionError as e:
        raise ConnectionError(f"Não foi possível conectar à Geranet: {str(e)[:200]}")
    except ValueError:
        raise
    except Exception as e:
        raise ConnectionError(f"Erro inesperado na consulta Geranet: {str(e)[:200]}")


def extrair_notas_mais_recentes(
    resposta: dict,
    limite: int = 10,
) -> list:
    """
    Extrai as notas mais recentes da resposta da Geranet.

    Args:
        resposta: Resposta completa da consultar_notas()
        limite: Número máximo de notas para retornar

    Returns:
        Lista de registros simplificados com dados principais
    """
    registros = resposta.get("registros", [])
    if not registros:
        return []

    # Ordena por data decrescente
    registros.sort(key=lambda r: r.get("data", ""), reverse=True)

    simplificadas = []
    for reg in registros[:limite]:
        dados_nota = reg.get("dadosNota", {})
        servico = dados_nota.get("servico", {})
        valores = servico.get("valores", {})
        tomador = dados_nota.get("tomador", {})

        simplificadas.append({
            "nsu": reg.get("nsu"),
            "numero_nota": reg.get("numeroNota"),
            "chave": reg.get("chaveDfe"),
            "codigo_verificacao": reg.get("codigoVerificacao"),
            "situacao": reg.get("descricaoSituacao"),
            "data_emissao": dados_nota.get("identificacao", {}).get("dataEmissao", reg.get("data")),
            "valor": valores.get("valorServicos", 0),
            "valor_liquido": valores.get("valorLiquidoNfse", 0),
            "tomador_nome": tomador.get("razaoSocial", ""),
            "tomador_cnpj": tomador.get("cpfCnpj", ""),
            "item_servico": servico.get("itemListaServico", ""),
            "discriminacao": servico.get("discriminacao", ""),
            "iss_retido": servico.get("tributacao", {}).get("issRetido", 0),
            "link": reg.get("link", ""),
            "tem_xml": bool(reg.get("xml")),
        })

    return simplificadas


def resumir_consulta(resposta: dict) -> dict:
    """
    Gera um resumo amigável da consulta para exibição ou contexto LLM.

    Args:
        resposta: Resposta completa da consultar_notas()

    Returns:
        Dict com resumo
    """
    registros = resposta.get("registros", [])
    notas = extrair_notas_mais_recentes(resposta, limite=20)

    total_valor = sum(n.get("valor", 0) or 0 for n in notas)
    notas_ano = [
        n for n in notas
        if n.get("data_emissao") and n["data_emissao"][:4] == "2026"
    ]
    total_ano = sum(n.get("valor", 0) or 0 for n in notas_ano)
    retidos = [n for n in notas if n.get("iss_retido") == 1]

    return {
        "quantidade_total": resposta.get("quantidadeDocumentos", len(registros)),
        "quantidade_retornada": len(registros),
        "ultimo_nsu": resposta.get("ultimoNsu"),
        "proximo_nsu_sugerido": resposta.get("proximoNsuSugerido"),
        "tem_mais": resposta.get("temMaisRegistrosProvavelmente") == "sim",
        "notas_analisadas": len(notas),
        "total_bruto": total_valor,
        "total_2026": total_ano,
        "notas_com_iss_retido": len(retidos),
        "primeira_nota": notas[0] if notas else None,
        "ultima_nota": notas[-1] if notas else None,
    }

File: worker/src/geranet.ts
/**
 * Integración con la API Geranet para consulta de NFSe.
 * Equivalente a backend/geranet.py en Python.
 */

import { Env } from './config';

export interface GeranetConsultaParams {
  cnpj: string;
  inscripcion_municipal: string;
  razon_social: string;
  municipio: string;
  certificado_digital?: string;
  senha_certificado?: string;
  ultimo_nsu?: string;
  chave_nfse?: string;
}

export interface GeranetConsultaResult {
  status: 'sucesso' | 'error';
  datos?: Record<string, unknown>;
  resumen?: Record<string, unknown>;
  error?: string;
}

function obtenerCertificadoHex(env: Env, cnpj: string | null): string {
  // 1. JSON multi-cert (GERANET_CERTS_JSON)
  if (env.GERANET_CERTS_JSON && cnpj) {
    try {
      const certs = JSON.parse(env.GERANET_CERTS_JSON) as Record<string, string>;
      if (certs[cnpj]) {
        return certs[cnpj];
      }
    } catch {
      // ignorar
    }
  }

  // 2. Variable individual GERANET_CERT_{CNPJ}
  if (cnpj) {
    const varName = `GERANET_CERT_${cnpj}`;
    const valor = (env as Record<string, string | undefined>)[varName];
    if (valor) return valor;
  }

  // 3. Certificado único (legado)
  const certBase64 = (env as Record<string, string | undefined>).GERANET_CERT_BASE64;
  if (certBase64) return certBase64;

  throw new Error(
    cnpj
      ? `Ningún certificado encontrado para el CNPJ ${cnpj}. Configura GERANET_CERTS_JSON o GERANET_CERT_{CNPJ}.`
      : 'Ningún certificado configurado. Define GERANET_CERT_BASE64 o GERANET_CERTS_JSON.'
  );
}

export async function consultarNotas(
  params: GeranetConsultaParams,
  env: Env,
): Promise<GeranetConsultaResult> {
  if (!env.GERANET_API_KEY) {
    return { status: 'error', error: 'GERANET_API_KEY no configurada' };
  }

  let hexCert: string;
  try {
    hexCert = params.certificado_digital || obtenerCertificadoHex(env, params.cnpj);
  } catch (e) {
    return { status: 'error', error: e instanceof Error ? e.message : 'Error de certificado' };
  }

  const senha = params.senha_certificado || env.GERANET_CERT_PASSWORD;
  if (!senha) {
    return {
      status: 'error',
      error: 'Contraseña del certificado no informada. Envía senha_certificado o configura GERANET_CERT_PASSWORD.',
    };
  }

  const payload = {
    prestador: {
      cnpj: params.cnpj,
      inscripcionMunicipal: params.inscripcion_municipal,
      razonSocial: params.razon_social,
      municipio: params.municipio,
    },
    certificadoDigital: hexCert,
    senhaCertificadoDigital: senha,
    padraoNacional: 'sim',
    ultimoNsu: params.ultimo_nsu || '0',
    ...(params.chave_nfse ? { chaveNfse: params.chave_nfse } : {}),
  };

  try {
    const res = await fetch(`${env.GERANET_BASE_URL || 'https://nfe.geranet.net/api/v1'}/nfse/consultar-notas`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${env.GERANET_API_KEY}`,
      },
      body: JSON.stringify(payload),
    });

    if (res.status === 401) return { status: 'error', error: 'Credenciales Geranet inválidas (HTTP 401)' };
    if (res.status === 402) return { status: 'error', error: 'Saldo insuficiente en Geranet (HTTP 402)' };
    if (res.status >= 400) {
      const body = await res.text();
      return { status: 'error', error: `Geranet retornó HTTP ${res.status}: ${body.slice(0, 300)}` };
    }

    const data = await res.json();
    if (data.situacion === 'error') {
      return { status: 'error', error: `Geranet retornó error: ${data.mensaje || 'sin mensaje'}` };
    }

    return { status: 'sucesso', datos: data, resumen: resumirConsulta(data) };
  } catch (e) {
    return {
      status: 'error',
      error: `Error inesperado en la consulta Geranet: ${e instanceof Error ? e.message.slice(0, 200) : 'desconocido'}`,
    };
  }
}

export function extraerNotasMasRecientes(respuesta: Record<string, unknown>, limite = 10): Array<Record<string, unknown>> {
  const registros = (respuesta.registros as Array<Record<string, unknown>>) || [];
  if (registros.length === 0) return [];

  // Ordena por fecha descendente
  registros.sort((a, b) => String(b.data || '').localeCompare(String(a.data || '')));

  const simplificadas: Array<Record<string, unknown>> = [];
  for (const reg of registros.slice(0, limite)) {
    const datosNota = (reg.datosNota as Record<string, unknown>) || {};
    const servico = (datosNota.servico as Record<string, unknown>) || {};
    const valores = (servico.valores as Record<string, unknown>) || {};
    const tomador = (datosNota.tomador as Record<string, unknown>) || {};
    const identificacion = (datosNota.identificacion as Record<string, unknown>) || {};

    simplificadas.push({
      nsu: reg.nsu,
      numero_nota: reg.numeroNota,
      chave: reg.chaveDfe,
      codigo_verificacion: reg.codigoVerificacion,
      situacion: reg.descripcionSituacion,
      data_emision: identificacion.dataEmision || reg.data,
      valor: valores.valorServicios || 0,
      valor_liquido: valores.valorLiquidoNfse || 0,
      tomador_nombre: tomador.razonSocial || '',
      tomador_cnpj: tomador.cpfCnpj || '',
      item_servicio: servico.itemListaServicio || '',
      discriminacion: servico.discriminacion || '',
      iss_retido: (servico.tributacion as Record<string, unknown>)?.issRetido || 0,
      link: reg.link || '',
      tem_xml: Boolean(reg.xml),
    });
  }

  return simplificadas;
}

export function resumirConsulta(respuesta: Record<string, unknown>): Record<string, unknown> {
  const registros = (respuesta.registros as Array<Record<string, unknown>>) || [];
  const notas = extraerNotasMasRecientes(respuesta, 20);

  const totalValor = notas.reduce((acc, n) => acc + (Number(n.valor) || 0), 0);
  const notasAno = notas.filter((n) => n.data_emision && String(n.data_emision).startsWith('2026'));
  const totalAno = notasAno.reduce((acc, n) => acc + (Number(n.valor) || 0), 0);
  const retidos = notas.filter((n) => Number(n.iss_retido) === 1);

  return {
    cantidad_total: respuesta.quantidadeDocumentos || registros.length,
    cantidad_retornada: registros.length,
    ultimo_nsu: respuesta.ultimoNsu,
    proximo_nsu_sugerido: respuesta.proximoNsuSugerido,
    tem_mas: respuesta.temMasRegistrosProvavelmente === 'sim',
    notas_analizadas: notas.length,
    total_bruto: totalValor,
    total_2026: totalAno,
    notas_con_iss_retido: retidos.length,
    primera_nota: notas[0] || null,
    ultima_nota: notas[notas.length - 1] || null,
  };
}